server/autotest_server/utils.py

```diff
@@ -45,31 +43,54 @@ def _rlimit_str2int(rlimit_string):
     return getattr(resource, f"RLIMIT_{rlimit_string.upper()}")
 
 
-def set_rlimits_before_test() -> None:
+def validate_rlimit(config_soft: int, config_hard: int, curr_soft: int, curr_hard: int) -> tuple[int, int]:
+    """Validates and adjusts resource limits based on configured and current values.
+
+    Returns a tuple containing the validated (soft, hard) limit values. This implementation treats the current soft
+    limit as an upper bound on the config soft limit and will clamp it.
     """
-    Sets rlimit settings specified in config file
+    # account for the fact that resource.RLIM_INFINITY == -1
+    soft, hard = min(curr_soft, config_soft), min(curr_hard, config_hard)
+    if soft < 0:
+        soft = max(curr_soft, config_soft)
+    if hard < 0:
+        hard = max(curr_hard, config_hard)
+    # make sure the soft limit doesn't exceed the hard limit, but keep in mind that -1 is resource.RLIM_INFINITY
+    if hard != -1:
+        soft = min(hard, soft)
+
+    return soft, hard
+
+
+def get_resource_settings(config: _Config) -> list[tuple[int, tuple[int, int]]]:
+    """Returns rlimit settings specified in config file.
+
     This function ensures that for specific limits (defined in RLIMIT_ADJUSTMENTS),
     there are at least n=RLIMIT_ADJUSTMENTS[limit] resources available for cleanup
-    processes that are not available for test processes.  This ensures that cleanup
+    processes that are not available for test processes. This ensures that cleanup
     processes will always be able to run.
     """
-    for limit_str in config.get("rlimit_settings", {}).keys() | RLIMIT_ADJUSTMENTS.keys():
+    resource_settings = []
+
+    for limit_str in config.get("rlimit_settings", {}):
```

Comment:
Since we're now just iterating over one dictionary, you can tighten the code up a bit by using `.items()` here. There should always be a value provided for this value in the dictionary, so I don't think the `.get(limit_str, resource.getrlimit(limit)` on line 79 is actually necessary.